requests/structures.py

```diff
@@ -103,3 +104,94 @@ def __getitem__(self, key):
 
     def get(self, key, default=None):
         return self.__dict__.get(key, default)
+
+
+class TimedCache(object):
+    """
+    Evicts entries after expiration_secs.
+    If none are expired and maxlen is hit,
+    will evict the oldest cached entry
+    """
+
+    def __init__(self, maxlen=32, expiration_secs=60):
+        """
+        :param maxlen: most number of entries to hold on to
+        :param expiration_secs: the number of seconds to hold on
+        to entries
+        """
+
+        self.maxlen = maxlen
+        self.expiration_secs = expiration_secs
+        self._dict = dict()
+
+    def __repr__(self):
+        return '<timed-cache maxlen:%d len:%d>' % \
+            (self.maxlen, len(self._dict))
+
+    def __getitem__(self, key):
+        """
+        Look up an item in the cache. If the item
+        has already expired, it will be invalidated and not returned
+        :param key: which entry to look up
+        :return: the value in the cache, or None
+        """
+
+        return self.get(key, None)
```

Comment:
So, generally speaking it is better to implement `get` in terms of `__getitem__` than the other way around. `__getitem__` should almost always maintain dict-like behaviour of raising `KeyError` if the entry is not present, rather than returning `None`. The core concern here is that it needs to be possible to distinguish between "key not present in dictionary" and "key present with a value of `None`".